sample/weighted.go

```diff
@@ -0,0 +1,104 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import "math/rand"
+
+// Weighted is a collection of items that can be selected with weighted probabilities
+// without replacement.
+type Weighted struct {
+	indexes []int
+	weights []float64
+	totals  []float64
+	src     *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the indexes in idx weighted by w. Elements of
```

Comment:
I don't like the indexes argument. This type is for sampling from a weighted set of objects whose weights are represented by the w slice. This is fundamentally a cross between Intn() and Perm(), and like those methods should just return values between 0 and len(weight). The added semantics of being elements from an index slice are not relevant to the generality of the type. Why is it a []int as opposed to a []float64, []string or []optimize.Method? It seems to me that adding semantics for a particular kind of set are best used through embedding.

````
type EdgeSampler struct {
    sample.Weighted
    Ids []int
}

func (e EdgeSampler) Take() (id int, ok bool){
   idx, ok := e.Weighted.Take()
   if !ok {
       return -1, ok
   }
   return e.Ids[idx], true 
}